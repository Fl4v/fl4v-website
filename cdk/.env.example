CDK_PROD_ACCOUNT=account-id
CDK_DEV_ACCOUNT=account-id
