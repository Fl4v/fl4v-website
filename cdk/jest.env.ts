process.env.CDK_PROD_ACCOUNT = 'prod-account-id';
process.env.CDK_DEV_ACCOUNT = 'dev-account-id';
