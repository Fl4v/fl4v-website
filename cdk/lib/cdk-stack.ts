import { Stack, StackProps, Tags } from 'aws-cdk-lib';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';
import {
	Distribution,
	OriginAccessIdentity,
	PriceClass,
	ResponseHeadersPolicy,
	ViewerProtocolPolicy
} from 'aws-cdk-lib/aws-cloudfront';
import { S3Origin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, AaaaRecord, HostedZone, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket, BucketAccessControl } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export interface Fl4vStackProps extends StackProps {
	environment: string;
	domain: string;
	noindex: boolean;
}

export class CdkStack extends Stack {
	constructor(scope: Construct, id: string, props: Fl4vStackProps) {
		super(scope, id, props);

		Tags.of(this).add('service', props.domain);
		Tags.of(this).add('env', props.environment);

		const canonicalUser = new OriginAccessIdentity(this, 'access-identity');
		const noIndexPolicy = new ResponseHeadersPolicy(this, 'response-header-policy', {
			responseHeadersPolicyName: 'noindex-policy',
			customHeadersBehavior: {
				customHeaders: [{ header: 'X-Robots-Tag', value: 'noindex', override: true }]
			}
		});

		const certificate = new Certificate(this, 'certificate', {
			domainName: props.domain,
			certificateName: props.domain
		});

		const cloudfrontS3Bucket = new Bucket(this, 'cloudfront-s3-bucket', {
			accessControl: BucketAccessControl.PRIVATE,
			bucketName: props.domain
		});
		cloudfrontS3Bucket.grantRead(canonicalUser);

		const cloudFrontDistribution = new Distribution(this, props.domain, {
			defaultRootObject: 'index.html',
			defaultBehavior: {
				origin: new S3Origin(cloudfrontS3Bucket, {
					originAccessIdentity: canonicalUser
				}),
				viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
				responseHeadersPolicy: props.noindex ? noIndexPolicy : undefined
			},
			priceClass: PriceClass.PRICE_CLASS_100,
			domainNames: [props.domain],
			certificate: certificate,
			errorResponses: [
				{
					httpStatus: 404,
					responsePagePath: '/4xx-errors/404/'
				}
			]
		});

		const hostedZone = HostedZone.fromLookup(this, 'hosted-zone', {
			domainName: props.domain
		});

		new ARecord(this, 'a-record-root', {
			zone: hostedZone,
			target: RecordTarget.fromAlias(new CloudFrontTarget(cloudFrontDistribution)),
			recordName: props.domain
		});
		new AaaaRecord(this, 'aaaa-record-root', {
			zone: hostedZone,
			target: RecordTarget.fromAlias(new CloudFrontTarget(cloudFrontDistribution)),
			recordName: props.domain
		});
	}
}
