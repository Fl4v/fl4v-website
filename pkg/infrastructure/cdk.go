package main

import (
	"os"

	"github.com/aws/aws-cdk-go/awscdk"
	"github.com/aws/aws-cdk-go/awscdk/awslambda"
	"github.com/aws/aws-cdk-go/awscdk/awslambdago"
	"github.com/aws/constructs-go/constructs/v3"
	"github.com/aws/jsii-runtime-go"
)

type stackProps struct {
	awscdk.StackProps
}

func lambdaStack(scope constructs.Construct, id string, props *stackProps) awscdk.Stack {
	var sprops awscdk.StackProps
	if props != nil {
		sprops = props.StackProps
	}
	stack := awscdk.NewStack(scope, &id, &sprops)

	// AWS lambda function definition
	awslambdago.NewGoFunction(stack, jsii.String("fl4v-web"), &awslambdago.GoFunctionProps{
		FunctionName: jsii.String("fl4v-web"),
		Entry:        jsii.String("../lambda/lambda.go"),
		Runtime:      awslambda.Runtime_GO_1_X(),
	})

	return stack
}

func main() {
	app := awscdk.NewApp(nil)

	// Stack properties
	props := stackProps{
		awscdk.StackProps{
			Env: &awscdk.Environment{
				Account: jsii.String(os.Getenv("CDK_DEFAULT_ACCOUNT")),
				Region:  jsii.String(os.Getenv("CDK_DEFAULT_REGION")),
			},
		},
	}

	lambdaStack(app, "Fl4vWeb", &props)

	app.Synth(nil)
}
