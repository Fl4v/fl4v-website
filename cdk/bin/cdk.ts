import { App } from 'aws-cdk-lib';
import { config } from 'dotenv';
import { CdkStack, Fl4vStackProps } from '../lib/cdk-stack';

config();

const app = new App();

export const prodStackProperties: Fl4vStackProps = {
	environment: 'prod',
	env: {
		account: process.env.CDK_PROD_ACCOUNT,
		region: 'us-east-1'
	},
	domain: 'fl4v.com',
	noindex: true
};

new CdkStack(app, 'prod-fl4v', prodStackProperties);
