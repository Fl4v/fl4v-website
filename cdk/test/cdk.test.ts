import { App } from 'aws-cdk-lib';
import { CdkStack } from '../lib/cdk-stack';
import { prodStackProperties } from '../bin/cdk';
import { Template } from 'aws-cdk-lib/assertions';

describe('Stack snapshot tests', () => {
	test('Prod stack', () => {
		const stack = new CdkStack(new App(), 'fl4v-prod', prodStackProperties);
		const template = Template.fromStack(stack);

		expect(template).toMatchSnapshot();
	});
});
