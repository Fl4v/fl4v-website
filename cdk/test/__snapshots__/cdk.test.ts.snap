// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack snapshot tests Prod stack 1`] = `
{
  "Mappings": {
    "AWSCloudFrontPartitionHostedZoneIdMap": {
      "aws": {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "aaaarecordroot016CB348": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "fl4vcomB78BD089",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "fl4v.com.",
        "Type": "AAAA",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "accessidentity09D4AC8C": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "arecordroot61DC1D81": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "fl4vcomB78BD089",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "fl4v.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "certificateEC031123": {
      "Properties": {
        "DomainName": "fl4v.com",
        "DomainValidationOptions": [
          {
            "DomainName": "fl4v.com",
            "ValidationDomain": "fl4v.com",
          },
        ],
        "Tags": [
          {
            "Key": "env",
            "Value": "prod",
          },
          {
            "Key": "Name",
            "Value": "fl4v.com",
          },
          {
            "Key": "service",
            "Value": "fl4v.com",
          },
        ],
        "ValidationMethod": "EMAIL",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "cloudfronts3bucketCA3A8CDD": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccessControl": "Private",
        "BucketName": "fl4v.com",
        "Tags": [
          {
            "Key": "env",
            "Value": "prod",
          },
          {
            "Key": "service",
            "Value": "fl4v.com",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "cloudfronts3bucketPolicy468F7DA1": {
      "Properties": {
        "Bucket": {
          "Ref": "cloudfronts3bucketCA3A8CDD",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "accessidentity09D4AC8C",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "cloudfronts3bucketCA3A8CDD",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "cloudfronts3bucketCA3A8CDD",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "accessidentity09D4AC8C",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "cloudfronts3bucketCA3A8CDD",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "fl4vcomB78BD089": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "fl4v.com",
          ],
          "CustomErrorResponses": [
            {
              "ErrorCode": 404,
              "ResponseCode": 404,
              "ResponsePagePath": "/4xx-errors/404/",
            },
          ],
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "ResponseHeadersPolicyId": {
              "Ref": "responseheaderpolicy3D273F8A",
            },
            "TargetOriginId": "fl4vprodfl4vcomOrigin1C883DFEB",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "cloudfronts3bucketCA3A8CDD",
                  "RegionalDomainName",
                ],
              },
              "Id": "fl4vprodfl4vcomOrigin1C883DFEB",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "accessidentity09D4AC8C",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Ref": "certificateEC031123",
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
        "Tags": [
          {
            "Key": "env",
            "Value": "prod",
          },
          {
            "Key": "service",
            "Value": "fl4v.com",
          },
        ],
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "responseheaderpolicy3D273F8A": {
      "Properties": {
        "ResponseHeadersPolicyConfig": {
          "CustomHeadersConfig": {
            "Items": [
              {
                "Header": "X-Robots-Tag",
                "Override": true,
                "Value": "noindex",
              },
            ],
          },
          "Name": "noindex-policy",
        },
      },
      "Type": "AWS::CloudFront::ResponseHeadersPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
